CorsixTH/Lua/research_department.lua

```diff
@@ -505,39 +506,41 @@ function ResearchDepartment:discoverObject(object, automatic)
   self.research_progress[object].discovered = true
 
   -- Go through all rooms to see if another one can be made available.
-  for room, _ in pairs(self.hospital.undiscovered_rooms) do
-    local discovery = true
-    for needed, _ in pairs(room.objects_needed) do
-      local obj = self.research_progress[TheApp.objects[needed]]
-      if obj and not obj.discovered then
-        discovery = false
-        break
-      end
-    end
-    if discovery then
-      self.hospital.discovered_rooms[room] = true
-      self.hospital.undiscovered_rooms[room] = nil
-      if self.hospital == self.world.ui.hospital then
-        if automatic then
-          self.world.ui.adviser:say(_A.research.new_available:format(object.name))
-        else
-          self.world.ui.adviser:say(_A.research.new_machine_researched:format(object.name))
+  for i, room_id in ipairs(self.hospital.room_discovery) do
+    if not room_id.is_discovered then
+      local room = room_id.room
+      local discovery = true
+      for needed, _ in pairs(room.objects_needed) do
+        local obj = self.research_progress[TheApp.objects[needed]]
+        if obj and not obj.discovered then
+          discovery = false
+          break
         end
       end
-      -- It may now be possible to continue researching machine improvements
-      local current_improvement_research = self.research_policy.improvements.current
-      local max_strength = self.world.map.level_config.gbv.MaxObjectStrength
-      local min_strength = max_strength
-      -- If we're not researching any improvement right now, and the newest discovery was
-      -- a machine that requires an improvement, switch the current policy.
-      if (not current_improvement_research or current_improvement_research.dummy) then
-        for research_object, progress in pairs(self.research_progress) do
-          if research_object.default_strength then
-            -- Don't improve those that already have the max strength
-            if progress.start_strength < max_strength then
-              if progress.discovered and progress.start_strength < min_strength then
-                self.research_policy["improvements"].current = research_object
-                min_strength = progress.start_strength
+      if discovery then
+        self.hospital.room_discovery[i].is_discovered = true
+        if self.hospital == self.world.ui.hospital then
```

Comment:
There is a standard check for this (`isLocalPlayerHospital` iirc), please use that